core/src/main/scala/cats/Applicative.scala

```diff
@@ -125,17 +127,20 @@ trait Applicative[F[_]] extends Apply[F] with InvariantMonoidal[F] { self =>
     if (n <= 0) unit
     else {
       val fvoid = void(fa)
-      // invariant n >= 1
-      def loop(n: Int): F[Unit] =
-        if (n == 1) fvoid
-        else {
-          // since n >= 2, then (n >> 1) >= 1 so we can call loop
-          val half = loop(n >> 1)
-          val both = productR(half)(half)
-          if ((n & 1) == 1) productR(both)(fvoid)
-          else both
-        }
-      loop(n)
+
+      // invariant: n >= 1
+      @tailrec def loop(fa: F[Unit], n: Int, acc: F[Unit]): F[Unit] =
+        if (n == 1) productR(fa)(acc)
+        else
+          // n >= 2
+          // so (n >> 1) >= 1 and we are allowed to call loop
+          loop(
+            productR(fa)(fa),
+            n >> 1,
+            if ((n & 1) == 1) productR(acc)(fa) else acc
```

Comment:
same comment here about leveraging `productR(unit)(a) == a` to avoid doing a needless `productR(fa)(unit)`